### PR TITLE
Add support for derived facts

### PR DESCRIPTION
**Pull request by jeffmay**
- Adds support for facts that are derived from other fact expression results

Depends on:
- [x] https://github.com/jeffmay/vapors/pull/25

### CODE REVIEW

File: core/src/main/scala/com/rallyhealth/vapors/factfilter/data/Fact.scala [hunk 1]
```diff
@@ -10,18 +10,29 @@ sealed abstract class Fact {
   val typeInfo: FactType[Value]
   val value: Value
 
-  override def toString: String = {
-    // TODO: JSON format for value? Show for value?
-    s"Fact(${typeInfo.fullName} = $value)"
+  // TODO: JSON format for value? Show for value?
```
Comment (kionka-rally) [inline core/src/main/scala/com/rallyhealth/vapors/factfilter/data/Fact.scala]: `Show` should have been called `Debug` if that makes sense. Another typeclass should be used for JSON representations.

File: core/src/main/scala/com/rallyhealth/vapors/factfilter/data/Fact.scala [hunk 2]
```diff
@@ -10,18 +10,29 @@ sealed abstract class Fact {
...
+  override def toString: String = this match {
+    case derivedFact: DerivedFact =>
```
Comment (kionka-rally) [inline core/src/main/scala/com/rallyhealth/vapors/factfilter/data/Fact.scala]: Why are you using reflection here?

File: core/src/main/scala/com/rallyhealth/vapors/factfilter/data/Fact.scala [hunk 3]
```diff
@@ -54,15 +65,28 @@ object Fact {
...
 }
 
 object TypedFact {
 
+  final def apply[A](
+    typeInfo: FactType[A],
+    value: A,
+  ): TypedFact[A] = SourceFactOfType(typeInfo, value)
+
+  final def apply[A](
```
Comment (kionka-rally) [inline core/src/main/scala/com/rallyhealth/vapors/factfilter/data/Fact.scala]: I'm not a big fan of the method overloading here.

File: core/src/main/scala/com/rallyhealth/vapors/factfilter/data/Fact.scala [hunk 4]
```diff
@@ -80,3 +104,39 @@ object TypedFact {
...
+    case _ => None
+  }
+}
+
+final case class SourceFactOfType[A](
+  typeInfo: FactType[A],
+  value: A,
+) extends TypedFact[A]
+
+object SourceFactOfType {}
```
Comment (kionka-rally) [inline core/src/main/scala/com/rallyhealth/vapors/factfilter/data/Fact.scala]: What is the purpose of the empty companion objects?

File: core/src/main/scala/com/rallyhealth/vapors/factfilter/data/Fact.scala [hunk 5]
```diff
@@ -54,15 +65,28 @@ object Fact {
...
+
+  final def apply[A](
+    typeInfo: FactType[A],
+    value: A,
+    evidence: Evidence,
+  ): TypedFact[A] with DerivedFact = DerivedFactOfType(typeInfo, value, evidence)
+
+  final def unapply[A](fact: TypedFact[A]): Some[(FactType[A], A)] = fact match {
+    case SourceFactOfType(typeInfo, value) => Some((typeInfo, value))
+    case DerivedFactOfType(typeInfo, value, _) => Some((typeInfo, value))
```
Comment (kionka-rally) [inline core/src/main/scala/com/rallyhealth/vapors/factfilter/data/Fact.scala:73]: Does this lose important information?

File: core/src/main/scala/com/rallyhealth/vapors/factfilter/data/FactOrFactSet.scala [hunk 6]
```diff
@@ -0,0 +1,20 @@
+package com.rallyhealth.vapors.factfilter.data
+
+import cats.Order
+import cats.instances.order._
+
+import scala.collection.immutable.SortedSet
+
+final class FactOrFactSet private[FactOrFactSet] (val toSortedSet: SortedSet[Fact]) extends AnyVal
```
Comment (kionka-rally) [inline core/src/main/scala/com/rallyhealth/vapors/factfilter/data/FactOrFactSet.scala:8]: `FactOrFactSet` is probably asking for another name. Perhaps just `FactSet`?

File: core/src/main/scala/com/rallyhealth/vapors/factfilter/data/FactOrFactSet.scala [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+
+object FactOrFactSet {
+
+  implicit def setOfOneFact(fact: Fact)(implicit order: Order[Fact]): FactOrFactSet = new FactOrFactSet(SortedSet(fact))
```
Comment (kionka-rally) [inline core/src/main/scala/com/rallyhealth/vapors/factfilter/data/FactOrFactSet.scala:12]: These implicits seem a bit aggressive.

File: core/src/main/scala/com/rallyhealth/vapors/factfilter/data/Fact.scala (hunk 2 above)
Comment (jeffmay) [reply core/src/main/scala/com/rallyhealth/vapors/factfilter/data/Fact.scala]: Good point. It's a `sealed trait`, so I can just override toString in the case classes.

File: core/src/main/scala/com/rallyhealth/vapors/factfilter/evaluator/InterpretExprAsFunction.scala [hunk 8]
```diff
@@ -377,7 +393,10 @@ object InterpretExprAsFunction {
   final case class Output[R](
     value: R,
     evidence: Evidence,
-  )
+  ) {
+
+    def addEvidence(evidence: Evidence): Output[R] = copy(evidence = this.evidence ++ evidence)
```
Comment (jeffmay) [inline core/src/main/scala/com/rallyhealth/vapors/factfilter/evaluator/InterpretExprAsFunction.scala]: Is this still used?

File: core/src/main/scala/com/rallyhealth/vapors/factfilter/evaluator/InterpretExprAsFunction.scala [hunk 9]
```diff
@@ -252,18 +263,20 @@ final class InterpretExprAsFunction[F[_] : Foldable, V, P]
...
-    val postParam = expr.capture.foldToParam(expr, input, subResult.output, allParams)
-    ExprResult.UsingDefinitions(expr, ExprResult.Context(input, subResult.output, postParam), subResult)
+    val postParam = expr.capture.foldToParam(expr, input, output, allParams)
+    ExprResult.UsingDefinitions(expr, ExprResult.Context(input, output, postParam), subResult)
   }
 
   override def visitWhen[R](expr: Expr.When[F, V, R, P]): Input[F, V] => ExprResult[F, V, R, P] = { input =>
     val conditionResult = expr.condExpr.visit(this)(input)
+    // TODO: Should this include the original input.evidence?
     val subInput = input.copy(evidence = conditionResult.output.evidence)
```
Comment (jeffmay) [inline core/src/main/scala/com/rallyhealth/vapors/factfilter/evaluator/InterpretExprAsFunction.scala]: Probably not...

File: core/src/main/scala/com/rallyhealth/vapors/factfilter/evaluator/InterpretExprAsFunction.scala [hunk 10]
```diff
@@ -213,7 +224,7 @@ final class InterpretExprAsFunction[F[_] : Foldable, V, P]
   }
 
   override def visitReturnInput(expr: Expr.ReturnInput[F, V, P]): Input[F, V] => ExprResult[F, V, F[V], P] = { input =>
-    resultOfPureExpr(expr, input, input.value, input.evidence) {
+    resultOfPureExpr(expr, input, input.value, input.evidence ++ Evidence.fromAnyOrNone(input.value)) {
```
Comment (jeffmay) [inline core/src/main/scala/com/rallyhealth/vapors/factfilter/evaluator/InterpretExprAsFunction.scala:227]: I should document this. Basically, if the input is a Fact or an Iterable[Fact], then it is added to its own evidence when moved to the output value.

File: core/src/main/scala/com/rallyhealth/vapors/factfilter/evaluator/InterpretExprAsFunction.scala [hunk 11]
```diff
@@ -128,7 +135,9 @@ final class InterpretExprAsFunction[F[_] : Foldable, V, P]
     val inputResult = expr.inputExpr.visit(this)(input)
     val flatMapFn = expr.flatMapExpr.visit(InterpretExprAsFunction())
     val allResults = inputResult.output.value.map { elem =>
-      val flatMapInput = input.withValue(elem)
+      // If given a fact, use it as evidence, otherwise keep input evidence for this value of the collection
+      val inputEvidence = Evidence.fromAny(elem).getOrElse(inputResult.output.evidence)
```
Comment (jeffmay) [inline core/src/main/scala/com/rallyhealth/vapors/factfilter/evaluator/InterpretExprAsFunction.scala:139]: The comment is repeated 3 times... I should probably move it to helper code.

File: core/src/main/scala/com/rallyhealth/vapors/factfilter/data/Evidence.scala [hunk 12]
```diff
@@ -31,6 +32,25 @@ final class Evidence private (val factSet: SortedSet[Fact]) extends AnyVal {
     case _ => Evidence(this.factSet | that.factSet)
   }
 
+  // TODO: Use a separate type?
```
Comment (jeffmay) [inline core/src/main/scala/com/rallyhealth/vapors/factfilter/data/Evidence.scala]: Remove this TODO. I don't think it is helpful to have a separate type.

File: core/src/main/scala/com/rallyhealth/vapors/factfilter/data/FactOrFactSet.scala (hunk 6 above)
Comment (jeffmay) [reply core/src/main/scala/com/rallyhealth/vapors/factfilter/data/FactOrFactSet.scala:8]: There already is a `FactSet`. This is magnet type. Its only purpose is to avoid method overloading for the `Evidence` and `FactTable` factory methods.

File: core/src/main/scala/com/rallyhealth/vapors/factfilter/data/Fact.scala (hunk 5 above)
Comment (jeffmay) [reply core/src/main/scala/com/rallyhealth/vapors/factfilter/data/Fact.scala:73]: If the evidence field were important, the caller would use the `DerivedFact` extractor object.

File: core/src/main/scala/com/rallyhealth/vapors/factfilter/data/FactOrFactSet.scala (hunk 7 above)
Comment (jeffmay) [reply core/src/main/scala/com/rallyhealth/vapors/factfilter/data/FactOrFactSet.scala:12]: This is the main purpose for this `FactOrFactSet` type. It is the magnet pattern to avoid overloading the factory methods for `FactTable` and `Evidence`. It's syntactic sugar for consistent wrapping facts (or sets of facts) as FactTables or Evidence.

File: core/src/main/scala/com/rallyhealth/vapors/factfilter/dsl/ExprBuilder.scala [hunk 13]
```diff
@@ -37,10 +37,11 @@ object ExprBuilder extends ExprBuilderSyntax {
 
 trait ExprBuilderLowPriorityImplicits {
 
+  // TODO: Add unit test for this
   implicit def liftFoldableExpr[F[_] : Foldable, V, M[_] : Foldable, U, P](
```
Comment (jeffmay) [inline core/src/main/scala/com/rallyhealth/vapors/factfilter/dsl/ExprBuilder.scala]: I don't think this is used, and I'm not sure it even works. I should write a unit test to verify or remove this code.

File: core/src/main/scala/com/rallyhealth/vapors/factfilter/evaluator/InterpretExprAsFunction.scala [hunk 14]
```diff
@@ -340,17 +354,19 @@ object InterpretExprAsFunction {
 
   final case class Input[F[_], V](
     value: F[V],
-    evidence: Evidence,
+    evidence: Evidence, // TODO: Is this needed? Can't I just add the input value as evidence when it is made of Facts?
```
Comment (jeffmay) [inline core/src/main/scala/com/rallyhealth/vapors/factfilter/evaluator/InterpretExprAsFunction.scala]: This seems needed because it must be possible to select a value (like `0.8`) from a fact (like `ProbabilityToUse().scores("weightloss")`) without losing the fact it came from.

My hope was that I would be able to combine the evidence at a higher level in the expression once the value has been used, but I feel like there are too many edge cases to handle there and I'm not sure it is possible. However, with the current input.evidence, it is hard to know when to use it. I think there just needs to be a lot more unit tests around expected behavior before I'll know whether my hypothesis that we can handle this at some higher level expression can be proven or disproven.

Comment (jeffmay) [reply core/src/main/scala/com/rallyhealth/vapors/factfilter/evaluator/InterpretExprAsFunction.scala]: I'm gonna remove this comment for now as I am doubting its possibility again and it isn't relevant to this PR.

File: core/src/main/scala/com/rallyhealth/vapors/factfilter/evaluator/InterpretExprAsFunction.scala (hunk 11 above)
Comment (paul-a-kennedy-rally) [reply core/src/main/scala/com/rallyhealth/vapors/factfilter/evaluator/InterpretExprAsFunction.scala:139]: yeah...

**Comment by jeffmay** [comment]
Just as a note. I am planning to rework how evidence tracking is done. A lot of this work will have to be revised, but as an initial proof of concept implementation, I believe this is sufficient.

**Review by darmbrus — APPROVED** [review]

File: core/src/main/scala/com/rallyhealth/vapors/factfilter/data/FactOrFactSet.scala (hunk 6 above)
Comment (paul-a-kennedy-rally) [reply core/src/main/scala/com/rallyhealth/vapors/factfilter/data/FactOrFactSet.scala:8]: I agree that there really needs to be better names in this.  I know that your development is a bit of a hydra-model right now, so I don't know how painful this will be to do in the middle.

But this will need to be done before too much longer.

File: core/src/main/scala/com/rallyhealth/vapors/factfilter/evaluator/InterpretExprAsFunction.scala (hunk 10 above)
Comment (paul-a-kennedy-rally) [reply core/src/main/scala/com/rallyhealth/vapors/factfilter/evaluator/InterpretExprAsFunction.scala:227]: Yes, please do!

**Review by paul-a-kennedy-rally — APPROVED** [review]
There's a renaming armgeddon coming, but no immediate blockers.

File: core/src/main/scala/com/rallyhealth/vapors/factfilter/data/FactOrFactSet.scala (hunk 6 above)
Comment (jeffmay) [reply core/src/main/scala/com/rallyhealth/vapors/factfilter/data/FactOrFactSet.scala:8]: I think there may be some misunderstanding about this type. It should not be referenced directly. It only exists as a "magnet type", which is basically a target for implicit conversion that you can put on a method (can also be as varargs) that allows you to define various implicit conversions for different known types. This allows you to avoid method overloading, as well as allowing the implicit resolution system to prove type correctness before dropping the unnecessary type parameters (depending on what you keep in the magnet type). In my case, I'm just using this magnet type to allow calling the `Evidence.apply()` method with any variable combination of single facts, list of facts, set of facts, etc. without needing to define separate apply methods. In addition, this same magnet type is used in the `FactTable.apply` method to allow the same types of arguments to be passed.

File: core/src/main/scala/com/rallyhealth/vapors/factfilter/evaluator/InterpretExprAsFunction.scala (hunk 10 above)
Comment (jeffmay) [reply core/src/main/scala/com/rallyhealth/vapors/factfilter/evaluator/InterpretExprAsFunction.scala:227]: I think I'm gonna wait until after I rework the evidence tracking to document it, since it is fairly broken right now.
